Source/WebCore/rendering/svg/RenderSVGResourcePattern.cpp

```diff
@@ -0,0 +1,225 @@
+{
+    return SVGLengthContext::resolveRectangle(&patternElement, attributes.patternUnits(), objectBoundingBox, attributes.x(), attributes.y(), attributes.width(), attributes.height());
+}
+
+bool RenderSVGResourcePattern::buildTileImageTransform(const RenderElement& renderer, const PatternAttributes& attributes, const SVGPatternElement& patternElement, FloatRect& patternBoundaries, AffineTransform& tileImageTransform) const
+{
+    FloatRect objectBoundingBox = renderer.objectBoundingBox();
+    patternBoundaries = calculatePatternBoundaries(attributes, objectBoundingBox, patternElement);
+    if (patternBoundaries.width() <= 0 || patternBoundaries.height() <= 0)
+        return false;
+
+    AffineTransform viewBoxCTM = SVGFitToViewBox::viewBoxToViewTransform(attributes.viewBox(), attributes.preserveAspectRatio(), patternBoundaries.width(), patternBoundaries.height());
+
+    // Apply viewBox/objectBoundingBox transformations.
+    if (!viewBoxCTM.isIdentity())
+        tileImageTransform = viewBoxCTM;
+    else if (attributes.patternContentUnits() == SVGUnitTypes::SVG_UNIT_TYPE_OBJECTBOUNDINGBOX)
+        tileImageTransform.scale(objectBoundingBox.width(), objectBoundingBox.height());
+
+    return true;
+}
+
+RefPtr<ImageBuffer> RenderSVGResourcePattern::createTileImage(GraphicsContext& context, const FloatSize& size, const FloatSize& scale, const AffineTransform& tileImageTransform, const PatternAttributes& attributes) const
+{
+    // This is equivalent to making createImageBuffer() use roundedIntSize().
+    auto roundedUnscaledImageBufferSize = [](const FloatSize& size, const FloatSize& scale) -> FloatSize {
+        auto scaledSize = size * scale;
+        return size - (expandedIntSize(scaledSize) - roundedIntSize(scaledSize)) * (scaledSize - flooredIntSize(scaledSize)) / scale;
+    };
+
+    auto tileSize = roundedUnscaledImageBufferSize(size, scale);
+
+    // FIXME: Use createImageBuffer(rect, scale), delete the above calculations and fix 'tileImageTransform'
+    auto tileImage = context.createScaledImageBuffer(tileSize, scale);
+    if (!tileImage)
+        return nullptr;
+
+    GraphicsContext& tileImageContext = tileImage->context();
+
+    // Apply tile image transformations.
+    if (!tileImageTransform.isIdentity())
+        tileImageContext.concatCTM(tileImageTransform);
+
+    AffineTransform contentTransformation;
+    if (attributes.patternContentUnits() == SVGUnitTypes::SVG_UNIT_TYPE_OBJECTBOUNDINGBOX)
+        contentTransformation = tileImageTransform;
+
+    SVGHitTestCycleDetectionScope clippingScope(*this);
+
+    // Draw the content into the ImageBuffer.
+    for (auto& child : childrenOfType<SVGElement>(*attributes.patternContentElement())) {
+        if (!child.renderer())
+            continue;
+        SVGRenderingContext::renderSubtreeToContext(tileImageContext, *child.renderer(), contentTransformation);
```

Comment:
That is all legacy stuff, SVGRenderingContext does not work for LBSE. Transformations would be broken, rendering at wrong spots -- you need to paint/hit-test via layers, not using the render tree.

Excerpt from downstream RenderSVGResourcePattern:

```
    auto* patternRenderer = static_cast<RenderSVGResourcePattern*>(attributes.patternContentElement()->renderer());
    ASSERT(patternRenderer);
    ASSERT(patternRenderer->hasLayer());
    
    // Draw the content into the ImageBuffer.
    patternRenderer->layer()->paintSVGResourceLayer(tileImageContext, LayoutRect::infiniteRect(), tileImageTransform);
```